Sources/TSCBasic/FileSystem.swift

```diff
@@ -224,41 +234,78 @@ public extension FileSystem {
     func getFileInfo(_ path: AbsolutePath) throws -> FileInfo {
         fatalError("This file system currently doesn't support this method")
     }
+
+    func resolveUnicode(_ path: AbsolutePath) -> AbsolutePath {
+        return path
+    }
 }
 
 /// Concrete FileSystem implementation which communicates with the local file system.
 private class LocalFileSystem: FileSystem {
 
+    func resolveUnicode(_ path: AbsolutePath) -> AbsolutePath {
+        #if os(macOS)
+        // The macOS file system enforces NFD.
+        // It can find everything without any help.
+        return path
+        #else
+        if _exists(path, followSymlink: true) {
+            return path
+        } else {
+            // Search for any neighbours with Unicode‐equivalent names
+            // in order to be resilient against Unicode‐naïveté in the file system.
+            let parent = resolveUnicode(path.parentDirectory)
+            let neighbours = (try? _getDirectoryContents(parent)) ?? []
+            if let equivalent = neighbours.first(where: { $0 == path.basename }) {
+                // Return the extant equivalent.
+                return parent.appending(component: equivalent)
+            } else {
+                // Nothing found; return unaltered (but still using resolved parent).
+                return parent.appending(component: path.basename)
+            }
+        }
+        #endif
+    }
+
     func isExecutableFile(_ path: AbsolutePath) -> Bool {
+        let path = resolveUnicode(path)
         // Our semantics doesn't consider directories.
         return  (self.isFile(path) || self.isSymlink(path)) && FileManager.default.isExecutableFile(atPath: path.pathString)
     }
 
-    func exists(_ path: AbsolutePath, followSymlink: Bool) -> Bool {
+    /// This method is Unicode‐naïve.
```

Comment:
Good indication that an assert would be nice; at the very least it is documentation about what you mean by "naive", as in it assumes the caller handled such details.

Reply:
See the other comment about types and assertions.

It is `private`. The documentation was a brief note to other contributors about why it is distinct from the neighbouring method. I could elaborate it with more of what I said in the other comment if any of you think it would be helpful.